File: appinventor/components/src/com/google/appinventor/components/runtime/ComponentContainer.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2009-2011 Google, All Rights reserved
// Copyright 2011-2012 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.runtime;

import android.app.Activity;
import android.view.View;
import org.w3c.dom.Text;

/**
 * Components that can contain other components need to implement this
 * interface.
 *
 */
public interface ComponentContainer {
  /**
   * Returns the activity context (which can be retrieved from the root
   * container - aka the form).
   *
   * @return  activity context
   */
  Activity $context();

  /**
   * Returns the form that ultimately contains this container.
   *
   * @return  form
   */
  Form $form();

  /**
   * Adds a component to a container.
   *
   * <p/>After this method is finished executing, the given component's view
   * must have LayoutParams, even if the component cannot be added to the
   * container until later.
   *
   * @param component  component associated with view
   */
  void $add(AndroidViewComponent<? extends View> component);

  void setChildWidth(AndroidViewComponent<? extends View> component, int width);

  void setChildHeight(AndroidViewComponent<? extends View> component, int height);

  int Width();

  int Height();

}

File: appinventor/components-ios/AIComponentKit/ComponentContainer.swift
// -*- mode: swift; swift-mode:basic-offset: 2; -*-
// Copyright © 2016-2020 Massachusetts Institute of Technology, All rights reserved.

import Foundation

@objc public protocol ComponentContainer {
  var form: Form { get }
  var container: ComponentContainer { get }
  func add(_ component: ViewComponent)
  func setChildWidth(of component: ViewComponent, to width: Int32)
  func setChildHeight(of component: ViewComponent, to height: Int32)
  func isVisible(component: ViewComponent) -> Bool
  func setVisible(component: ViewComponent, to visibility: Bool)
  func isVisible() -> Bool
  var Width: Int32 { get }
  var Height: Int32 { get }
}
